src/ucp/core/ucp_worker.c

```diff
@@ -2948,21 +2949,31 @@ ucp_worker_do_keepalive_progress(ucp_worker_h worker)
     }
 
     if (ucs_unlikely(worker->keepalive.iter == &worker->all_eps)) {
+        ucs_assert(worker->keepalive.iter_begin == &worker->all_eps);
+
         ucp_worker_keepalive_next_ep(worker);
+        worker->keepalive.iter_begin = worker->keepalive.iter;
```

Comment:
IMO we need to check this inside the loop, or in ucp_worker_keepalive_current_ep(), maybe it's possible we will have to "skip" over &worker->all_eps during loop iteration?